crates/driver/src/domain/eth/mod.rs

```diff
@@ -197,6 +200,54 @@ impl TokenAmount {
     }
 }
 
+impl Sub<Self> for TokenAmount {
+    type Output = TokenAmount;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        self.0.sub(rhs.0).into()
+    }
+}
+
+impl num::CheckedSub for TokenAmount {
+    fn checked_sub(&self, other: &Self) -> Option<Self> {
+        self.0.checked_sub(other.0).map(Into::into)
+    }
+}
+
+impl Mul<Self> for TokenAmount {
+    type Output = TokenAmount;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        self.0.mul(rhs.0).into()
+    }
+}
+
+impl num::CheckedMul for TokenAmount {
+    fn checked_mul(&self, other: &Self) -> Option<Self> {
+        self.0.checked_mul(other.0).map(Into::into)
+    }
+}
+
+impl num::CheckedAdd for TokenAmount {
+    fn checked_add(&self, other: &Self) -> Option<Self> {
+        self.0.checked_add(other.0).map(Into::into)
+    }
+}
+
+impl Div<Self> for TokenAmount {
```

Comment:
Making it easier to create div by zero errors does not seem like a good idea. All the other unchecked operations are way less likely to happen (actually unchecked sub is also pretty risky).

Reply:
@MartinquaXD This `trait` is mandatory to be implemented for the `checked_div` trait. Because one is the super trait of the other. All the types which implements `checked_div` *must* implement `div` too.

Reply:
Interesting. I wasn't aware of that. Wouldn't it make more sense the other way around? 🤔 
In any case then I'd say this is fine.

Reply:
> Wouldn't it make more sense the other way around? 🤔
@MartinquaXD No, the reason is because `checked_div()` under the hood is something like:
```rust
pub fn checked_div(&self, other: &Self) -> Option<Self> {
   if other.is_zero() {
      return None;
   }
   Some(self.div(other))
}
 ```
 (it is dirty code just for explanation).
 So it uses `div`, that's the reason it is forcing its implementation.

Reply:
Sure, but there are 2 ways to skin this cat. You could also do it the other way around which would conceptually make more sense to me. Specifically because of the case we are currently talking about:
```rust
fn div(&self, other: &Self) -> Self {
    self.checked_div(other).unwrap()
}
```
But I guess it was done the way it was done because `Div` is already built-in for the primitive types in the standard library so `CheckedDiv` can easily extend those. Also you might not want to rely on the compiler optimizing the unnecessary checks away when doing `.unwrap()`.
If you don't think about how things are implemented under the hood and only think about which type should extend which other type it makes more sense to me to have it the other way around. In fact this is also what they did in the standard library for `Ord` and `PartialOrd`.
`Ord` (which neglects edge cases) extends `PartialOrd` (which is aware of these edge cases).

Reply:
I totally understand your point of view and what you mean. But the guys who wrote those traits in Rust did it like that (I believe, I may be wrong) because if you do this:

```rust
fn div(&self, other: &Self) -> Self {
    self.checked_div(other).unwrap()
}
```
Then if it panics you miss the reason because of the `unwrap()`, whereas the other way around (how it is really implemented), the code fails because of a real panic ([with a message](https://doc.rust-lang.org/std/ops/trait.Div.html)). You can argue that you can do `.expect()` with a custom message, but then you lose the granularity, because the code which is called may panic for different reasons. And if you want to keep the original panic message then you need to add boilerplate code which can be avoided by not overriding the panic.